zcash_keys/src/keys/transparent.rs

```diff
@@ -0,0 +1,340 @@
+            //        secp256k1_ec_pubkey_serialize(ctx, ptr, &pubkeylen, &pubkey, SECP256K1_EC_UNCOMPRESSED);
+            //        ptr += pubkeylen;
+            //        *seckeylen = ptr - seckey;
+            //        assert(*seckeylen == CKey::PRIVATE_KEY_SIZE);
+            SecretVec::new(
+                begin
+                    .iter()
+                    .chain(keypair.secret_bytes().iter())
+                    .chain(middle.iter())
+                    .chain(keypair.public_key().serialize_uncompressed().iter())
+                    .copied()
+                    .collect(),
+            )
+        }
+    }
+
+    // Decodes a secret key from the "openssl-inspired" DER encoding used by zcashd.
+    pub fn der_decode(encoded: &SecretVec<u8>, compressed: bool) -> Result<Self, ()> {
+        // Ported from https://github.com/zcash/zcash/blob/1f1f7a385adc048154e7f25a3a0de76f3658ca09/src/key.cpp#L36
+        // The original c++ code is retained as comments.
+
+        let seckey = encoded.expose_secret();
+
+        //    /* sequence header */
+        //    if (end - seckey < 1 || *seckey != 0x30u) {
+        //        return 0;
+        //    }
+        //    seckey++;
+        if seckey.is_empty() || seckey[0] != 0x30 {
+            return Err(());
+        }
+        let seckey = &seckey[1..];
+
+        //    /* sequence length constructor */
+        //    if (end - seckey < 1 || !(*seckey & 0x80u)) {
+        //        return 0;
+        //    }
+        //    size_t lenb = *seckey & ~0x80u; seckey++;
+        //    if (lenb < 1 || lenb > 2) {
+        //        return 0;
+        //    }
+        //    if (end - seckey < lenb) {
+        //        return 0;
+        //    }
+        if seckey.is_empty() || (seckey[0] & 0x80 == 0) {
+            return Err(());
+        }
+        let lenb = usize::from(seckey[0] & !0x80);
+        let seckey = &seckey[1..];
+        if !(1..=2).contains(&lenb) {
+            return Err(());
+        }
+        if seckey.len() < lenb {
+            return Err(());
+        }
+
+        //    /* sequence length */
+        //    size_t len = seckey[lenb-1] | (lenb > 1 ? seckey[lenb-2] << 8 : 0u);
+        //    seckey += lenb;
+        //    if (end - seckey < len) {
+        //        return 0;
+        //    }
+        let len_low_bits = usize::from(seckey[lenb - 1]);
+        let len_high_bits = if lenb > 1 {
+            usize::from(seckey[lenb - 2]) << 8
+        } else {
+            0
+        };
+        let len = len_low_bits | len_high_bits;
+        let seckey = &seckey[lenb..];
+        if seckey.len() < len {
+            return Err(());
+        }
+
+        //    /* sequence element 0: version number (=1) */
+        //    if (end - seckey < 3 || seckey[0] != 0x02u || seckey[1] != 0x01u || seckey[2] != 0x01u) {
+        //        return 0;
+        //    }
+        //    seckey += 3;
+        if seckey.len() < 3 || seckey[0] != 0x02 || seckey[1] != 0x01 || seckey[2] != 0x01 {
```

Comment:
This is much clearer while still being close to the C++ code:
```suggestion
        if seckey.len() < 3 || &seckey[..3] != &[0x02, 0x01, 0x01] {
```

Non-blocking: This could use `slice::split_at` (similar to the earlier alternative).